Allow deleting all rows of a table using a new `simple_truncate` storage method

anoadragon453 opened the pull request:
This PR adds a new storage method called `simple_truncate` that does what it says on the tin: truncates a given database table.

In addition, this PR modifies a couple storage methods:

* Passing an empty dictionary for the `keyvalues` argument to `simple_delete` would result in an invalid query error. It now simply returns 0, for 0 rows modified.
* The `get_users` storage method has been deleted (it would return a list of *all* users on the homeserver) which was not in use. My IDE picked up on it while I was in the area.

This change was motivated by needing to clear all rows from a table as part of custom room preset work. Note that due to `iterable` being a required argument for `simple_delete_many`, it was not possible to select all rows for deletion.

anoadragon453 commented on synapse/storage/database.py:
```diff
@@ -2195,8 +2201,8 @@ def simple_delete_many_txn(
             values.append(value)
 
-        if clauses:
-            sql = "%s WHERE %s" % (sql, " AND ".join(clauses))
+        sql = "%s WHERE %s" % (sql, " AND ".join(clauses))
+
```
`clauses` is initialised to a non-falsey value right above, making the `if clauses:` conditional pointless.

DMRobertson commented:
Comment in passing: I think I'd prefer a new method with a sufficiently alarming name, like `delete_all_rows_from_table`. Others may disagree.

anoadragon453 commented:
Or perhaps another argument called `allow_dropping_all_rows`.

clokep commented:
> Or perhaps another argument called `allow_drop_all_rows`.

I'd much prefer simpler methods that do a single thing instead of a method with many parameters!

clokep commented:
Maybe it would make sense to provide a `truncate` function to do this using `TRUNCATE`?

anoadragon453 commented:
@clokep That could work yes, though note that the sql would need to be `DELETE FROM table_name` in SQLite, which does not support truncate: https://www.tutorialspoint.com/sqlite/sqlite_truncate_table.htm

clokep commented:
> @clokep That could work yes, though note that the sql would need to be `DELETE FROM table_name` in SQLite, which does not support truncate: [tutorialspoint.com/sqlite/sqlite_truncate_table.htm](https://www.tutorialspoint.com/sqlite/sqlite_truncate_table.htm)

Right, well seems like a reason to have a separate method still! (And probably make it so that the other methods *don't* support that?)

anoadragon453 commented:
I've reverted the changes to `simple_delete{,_many}` and added a `simple_truncate` method instead - plus a test! I agree that this is a less footgun-ny approach.

DMRobertson commented on synapse/storage/database.py:
```diff
@@ -2112,11 +2112,15 @@ def simple_delete_txn(
         Args:
             table: string giving the table name
-            keyvalues: dict of column names and values to select the row with
+            keyvalues: dict of column names and values to select the row with. If empty,
+                no rows will be deleted.
 
         Returns:
             The number of deleted rows.
         """
+        if not keyvalues:
+            return 0
```
Sanity check: do we ever call this method with an empty set of key values?

I wonder if this ought to raise some kind of exception telling the author to use `simple_truncate`?

anoadragon453 replied:
It looks like we don't currently, but considering it's a no-op, it would likely aid in debugging if this did fail loudly. Good suggestion! I'll have it raise a `ValueError` and update the docstring.

anoadragon453 replied:
Done in 90128dd35.

DMRobertson commented on synapse/storage/database.py:
```diff
@@ -2195,12 +2199,49 @@ def simple_delete_many_txn(
 
         return txn.rowcount
 
+    async def simple_truncate(self, table: str, desc: str) -> None:
```
Does the word `simple` actually do anything here? (What would a complex truncate look like?)

anoadragon453 replied:
Hmm, I'm not sure! I guess it saves you needing to write the SQL yourself?

There's no other operational methods in this class without `simple_`, so we could either start a new trend or keep on with the consistency.

As someone who's worked with Synapse quite a bit, I'd *expect* to see `db_pool.simple_truncate`, and often `db_pool.simpl<tab>` for possibilities. But I'm also obviously biased.

DMRobertson commented on changelog.d/13446.removal:
```diff
@@ -0,0 +1 @@
+Remove the unused `get_users` storage method.
```
Not sure this is worth calling out. I don't think this is user-visible, either to end users or module api users(?).

DMRobertson reviewed (commented):
I think the PR title (description?) needs updating.

I've left a few thoughts.

erikjohnston commented on synapse/storage/database.py:
```diff
@@ -2195,12 +2206,49 @@ def simple_delete_many_txn(
+            table: The name of the table to delete all rows from.
+        """
+        if isinstance(txn.database_engine, PostgresEngine):
+            sql = "TRUNCATE %s" % table
+        else:
+            # SQLite does not support the TRUNCATE command
+            sql = "DELETE FROM %s" % table
```
Please either use `% (table,)` or f-strings :slightly_smiling_face:

erikjohnston commented:
What's the rationale for not just allowing keyvalues to be empty to truncate? That matches what `simple_select` and co. do I think? I suppose it runs the risk of *accidentally* deleting all rows?

anoadragon453 commented:
> What's the rationale for not just allowing keyvalues to be empty to truncate? That matches what `simple_select` and co. do I think? I suppose it runs the risk of *accidentally* deleting all rows?

@erikjohnston Mmm, see [above comments](https://github.com/matrix-org/synapse/pull/13446#issuecomment-1204046544). It was primarily for reducing the risk of accidentally dropping a table, yes.

DMRobertson reviewed (approved):
I'm happy!

anoadragon453 commented:
So after all this, this may not be necessary for the Configurable Room Defaults work. I think I've found a design that doesn't involve using the database. I still need to prove it in implementation though.

Given as nothing else would use this currently, I'm going to close this PR. Sorry for the noise. If someone else wants this functionality, this PR is pretty much ready to go.
